### PR TITLE
Feat/2006 public routes 

### PR DESCRIPTION
**Pull request by abrl91**
Issue Number: #2006 

## PR Details

still in process:
what has already been done?
- handle the client-side
- add Public to the `EnumEntityPermissionType` model and Prisma Enum
- check isPublic on default guard in order to determine if the route should be public or not
- add logic for public permission of the entity, on every action - only for the controller (resolver on a different PR).

what is still missing?
- if all the methods in controller/resolver were set to public - remove unused imports such as AclFilterResposeInterceptor and AclValidateRequestInterceptor

 Important To Do:
 - open a pr from release 0.13 and change there the relevant Enum on Prisma schema and models

## PR Checklist
- [x] Tests for the changes have been added
- [x] `npm test` doesn't throw any error

### CODE REVIEW

File: packages/amplication-data-service-generator/src/util/ast.ts
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/util/ast.ts:808]: We better use a more specific version of this function instead of this generic version.
I think we need to create a function named "removeIdentifierFromUseInterceptorDecorator" in "nestjs-code-generation.ts" 

Different decorators use different constructors and accept different parameters, so this version will not fit other use cases. 

Also, you are expecting a format of a @common.useInterceptor and any other use case where the decorator will not use a namespace, it will not work - so specific version of the function will resolve this issue

File: packages/amplication-data-service-generator/src/util/set-endpoint-permission.ts
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/util/set-endpoint-permission.ts:35]: use const or Identifier

File: packages/amplication-data-service-generator/src/server/resource/controller/create-controller.ts
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/resource/controller/create-controller.ts:222]: I think this part should be included in `createToManyRelationMethods` - it will save another iteration on all the relation fields and creation of all the identifiers

File: packages/amplication-data-service-generator/src/util/ast.ts
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/util/ast.ts:783]: why do we still need this function?

File: packages/amplication-data-service-generator/src/util/nestjs-code-generation.ts
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/util/nestjs-code-generation.ts:43]: the decorator name should be const in the function and not a parameter

File: packages/amplication-data-service-generator/src/util/nestjs-code-generation.ts
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/util/nestjs-code-generation.ts:44]: change to NamedTyped.Identifier

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]
see comments

File: packages/amplication-data-service-generator/src/util/ast.ts
Comment (abrl91) [reply packages/amplication-data-service-generator/src/util/ast.ts:783]: Maybe I got you wrong. Your comment was something like "create a specific version of this function to interceptors", so I thought you mean to leave as a function for decorator without namespace and  create a new specific one for interceptors

File: packages/amplication-data-service-generator/src/util/nestjs-code-generation.ts
Comment (abrl91) [reply packages/amplication-data-service-generator/src/util/nestjs-code-generation.ts:43]: why? UseInterceptors() can get arguments (identifier[]) and you want that the caller of this function sends the argument that needs to be removed

File: packages/amplication-data-service-generator/src/util/ast.ts
Comment (abrl91) [reply packages/amplication-data-service-generator/src/util/ast.ts:783]: removeDecoratorByName is used to remove UseRoles decorator

File: packages/amplication-data-service-generator/src/util/nestjs-code-generation.ts
Comment (abrl91) [reply packages/amplication-data-service-generator/src/util/nestjs-code-generation.ts:44]: when I call this function I send the decorator name as a hardcoded string and in the if statement on the function above we have the following check:
```
if (callee.value && callee.value.property?.name === decoratorName
```
and `name` on callee.value.property is a string too

File: packages/amplication-data-service-generator/src/server/resource/controller/create-controller.ts
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/resource/controller/create-controller.ts:310]: why are you using Map here and object in the parent use cases? 
if you decide to go with Map, please use Map< EnumEntityAction, namedTypes.Identifier>
if you decide to use an object, you should also consider using a type for the object instead of using `as`

File: packages/amplication-data-service-generator/src/util/ast.ts
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/util/ast.ts:783]:
```suggestion
// This function removes all instances of a decorator by its name. in case the same decorator exists multiple times, it will be removed from all locations
export function removeDecoratorByName(
```

File: packages/amplication-data-service-generator/src/util/ast.ts
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/util/ast.ts:824]: are we still using this function? if not, please remove it.

File: packages/amplication-data-service-generator/src/util/nestjs-code-generation.ts
Comment (yuval-hazaz) [reply packages/amplication-data-service-generator/src/util/nestjs-code-generation.ts:43]: I am not talking about the identifier, I am talking about the decorator name and a not about the identifier

File: packages/amplication-data-service-generator/src/util/nestjs-code-generation.ts
Comment (yuval-hazaz) [reply packages/amplication-data-service-generator/src/util/nestjs-code-generation.ts:44]: in this case, I am talking about the identifier and not about the name of the decorator

File: packages/amplication-data-service-generator/src/util/set-endpoint-permission.ts
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/util/set-endpoint-permission.ts:22]: I dont think this is what you meant to write... 
if you are using entity.permissions.find() then I would expect you just to find the element and exit, and only then use the element you found to continue with the logic

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]

File: packages/amplication-data-service-generator/src/util/ast.ts
Comment (abrl91) [reply packages/amplication-data-service-generator/src/util/ast.ts:785]: added the comment

File: packages/amplication-data-service-generator/src/server/resource/resolver/resolver.base.template.ts
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/resource/resolver/resolver.base.template.ts:19]: since we are adding this conditionally from the code, should we remove this?

**Review by yuval-hazaz — APPROVED** [review]
@abrl91 great work!!
I approved all changes, but I think we should update the snapshot test to include at least one entity with public endpoints so we can see the impact on the snapshot. 
It will give us more confidence that the results are as expected.

File: packages/amplication-data-service-generator/src/server/resource/resolver/resolver.base.template.ts
Comment (abrl91) [reply packages/amplication-data-service-generator/src/server/resource/resolver/resolver.base.template.ts:19]: @yuval-hazaz the implementation of adding an import to 'Public' decorator is only in the create-controller for now. I decided to wait until you approve it and then merge it and use it in the create-resolver file. I will add the implementation now and remove this line

**Comment by abrl91** [comment]
> @abrl91 great work!! I approved all changes, but I think we should update the snapshot test to include at least one entity with public endpoints so we can see the impact on the snapshot. It will give us more confidence that the results are as expected.

@yuval-hazaz 
OK. I think we also need to check if all the methods in the controller or the resolver are public and if so, remove the imports of AclFilterResponseInreceptor and AclValidateRequestInterceptor.

**Comment by mshidlov** [comment]
@abrl91 lets merge this and not wait for that cleanup. we can work on it separately

**Comment by yuval-hazaz** [comment]
@abrl91 one last thing before merging - I think we should test all actions and not just view and search.

**Pull request merged**
